Fragment Directive not removed from URL.

megangardner opened the pull request:
#### c9e0352418fc0b8e92ed7e2f80370a38b75c148f
<pre>
Fragment Directive not removed from URL.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244318">https://bugs.webkit.org/show_bug.cgi?id=244318</a>

Reviewed by Ryosuke Niwa and Alex Christensen.

We were not stripping the fragment directive from the URL when setting the URL on the document.
This is necessary to be spec compliant as well as we would like to maintain user privacy and not
allow Text Fragment to become a tracking vector. Adding the WPT that now passes since all the
URL is stripped out, which was not the main thing that was being tested, but it now completely
passes where other tests still have failures I&apos;m working through.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc.html: Added.
* Source/WTF/wtf/URL.cpp:
(WTF::URL::consumefragmentDirective):
* Source/WTF/wtf/URL.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:
(WebCore::Document::fragmentDirective const):
* Source/WebCore/dom/FragmentDirectiveParser.cpp:
(WebCore::FragmentDirectiveParser::FragmentDirectiveParser):
* Source/WebCore/dom/FragmentDirectiveParser.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFragment):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getTextFragmentMatch):

Canonical link: <a href="https://commits.webkit.org/253802@main">https://commits.webkit.org/253802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a1570faf823a06df902600b1eecd129381d43de

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87131 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31218 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/17970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/96125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/149705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29583 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92747 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73945 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/17970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79003 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27308 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/17970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72674 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27255 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/17970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25899 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2682 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28939 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75484 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28879 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/17970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16710 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc.html:
```diff
@@ -0,0 +1,66 @@
+    });
+  }).then(position => {
+    assert_equals(position, 'text');
+    assert_equals(window.location.href.indexOf(':~:'), -1, 'Expected fragment directive to be stripped from the URL.');
```
You can use includes:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes

Maybe you should also test for "text="

megangardner replied:
I'll consider these as comments on the WPTs

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc.html:
```diff
@@ -0,0 +1,66 @@
+}
+
+function runTest() {
+  promise_test(t => new Promise(resolve => {
```
Use async function?
promise_test(async t => {
    await new Promise(requestAnimationFrame);
...

rniwa commented on Source/WTF/wtf/URL.h:
```diff
@@ -114,7 +114,6 @@ class URL {
     WTF_EXPORT_PRIVATE StringView lastPathComponent() const;
     WTF_EXPORT_PRIVATE StringView query() const;
     WTF_EXPORT_PRIVATE StringView fragmentIdentifier() const;
-
```
This seems unnecessary removal of a blank line?

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7575,7 +7575,12 @@ void WebPage::getTextFragmentMatch(CompletionHandler<void(const String&)>&& comp
     }
 
-    FragmentDirectiveParser fragmentDirectiveParser(mainWebFrame().url());
+    auto fragmentDirective = document->fragmentDirective();
```
Don't we need to make the same change in WebKitLegacy?

megangardner replied:
I'll consider this for a later change.

rniwa commented on Source/WebCore/dom/Document.h:
```diff
@@ -2273,7 +2273,7 @@ class Document
     UniqueRef<FrameSelection> m_selection;
         
-    String m_fragmentDirective;
+    std::optional<String> m_fragmentDirective;
```
Why don't we use nullAtom as null? That's distinguishable from an empty string.

rniwa reviewed (approved):
r=me

achristensen07 commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -238,6 +238,27 @@ StringView URL::fragmentIdentifier() const
+    auto fragment = fragmentIdentifier();
+    
+    if (fragment.isEmpty())
+        return nullString();
```
I often just write return { } unless I need to return a const String& in which case you need something like nullString()

achristensen07 commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -238,6 +238,27 @@ StringView URL::fragmentIdentifier() const
+    ASCIILiteral fragmentDirectiveDelimiter = ":~:"_s;
+    auto fragment = fragmentIdentifier();
+    
+    if (fragment.isEmpty())
```
I don't think this early return does anything.  If the fragment is empty, then find will return notFound.

achristensen07 commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -238,6 +238,27 @@ StringView URL::fragmentIdentifier() const
+    if (fragment.isEmpty())
+        return nullString();
+    
+    auto fragmentDirectiveStart = fragment.find(StringView(fragmentDirectiveDelimiter));
```
Calling the StringView constructor here isn't necessary.  StringView has a find that takes an ASCIILiteral.

achristensen07 commented on Source/WebCore/dom/Document.h:
```diff
@@ -1666,7 +1666,7 @@ class Document
         
     void setFragmentDirective(const String& fragmentDirective) { m_fragmentDirective = fragmentDirective; }
-    const String& fragmentDirective() const { return m_fragmentDirective; }
+    const String fragmentDirective() const { return m_fragmentDirective; }
```
I don't think this is a good change.  We only need to return a reference in a const getter.

megangardner committed:
Fragment Directive not removed from URL.
https://bugs.webkit.org/show_bug.cgi?id=244318

Reviewed by Ryosuke Niwa and Alex Christensen.

We were not stripping the fragment directive from the URL when setting the URL on the document.
This is necessary to be spec compliant as well as we would like to maintain user privacy and not
allow Text Fragment to become a tracking vector. Adding the WPT that now passes since all the
URL is stripped out, which was not the main thing that was being tested, but it now completely
passes where other tests still have failures I'm working through.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc.html: Added.
* Source/WTF/wtf/URL.cpp:
(WTF::URL::consumefragmentDirective):
* Source/WTF/wtf/URL.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:
(WebCore::Document::fragmentDirective const):
* Source/WebCore/dom/FragmentDirectiveParser.cpp:
(WebCore::FragmentDirectiveParser::FragmentDirectiveParser):
* Source/WebCore/dom/FragmentDirectiveParser.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFragment):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getTextFragmentMatch):

Canonical link: https://commits.webkit.org/253802@main
